src/backend/distributed/utils/citus_stat_tenants.c

```diff
@@ -652,45 +710,94 @@ CreateTenantStats(MultiTenantMonitor *monitor, TimestampTz queryTime)
 	 */
 	EvictTenantsIfNecessary(queryTime);
 
-	int tenantIndex = monitor->tenantCount;
+	bool found;
+	TenantStatsHashKey *key = CreateTenantStatsHashKey(AttributeToTenant,
+													   AttributeToColocationGroupId);
+
+	TenantStats *stats = (TenantStats *) hash_search(monitor->tenants, key,
+													 HASH_ENTER, &found);
+
+	pfree(key);
+
+	stats->writesInLastPeriod = 0;
+	stats->writesInThisPeriod = 0;
+	stats->readsInLastPeriod = 0;
+	stats->readsInThisPeriod = 0;
+	stats->cpuUsageInLastPeriod = 0;
+	stats->cpuUsageInThisPeriod = 0;
+	stats->score = 0;
+	stats->lastScoreReduction = 0;
+
+	if (!found)
+	{
+		/* initialize the stats lock for the new entry in the hash table */
+		SpinLockInit(&stats->lock);
+	}
+
+	return stats;
+}
+
+
+/*
+ * FindTenantStats finds the current tenant's statistics.
+ */
+static TenantStats *
+FindTenantStats(MultiTenantMonitor *monitor)
+{
+	TenantStatsHashKey *key = CreateTenantStatsHashKey(AttributeToTenant,
+													   AttributeToColocationGroupId);
+
+	TenantStats *stats = (TenantStats *) hash_search(monitor->tenants, key,
+													 HASH_FIND, NULL);
+
+	pfree(key);
+
+	return stats;
+}
 
-	memset(&monitor->tenants[tenantIndex], 0, sizeof(monitor->tenants[tenantIndex]));
 
-	strcpy_s(monitor->tenants[tenantIndex].tenantAttribute,
-			 sizeof(monitor->tenants[tenantIndex].tenantAttribute), AttributeToTenant);
-	monitor->tenants[tenantIndex].colocationGroupId = AttributeToColocationGroupId;
+static TenantStatsHashKey *
+CreateTenantStatsHashKey(char *tenantAttribute, uint32 colocationGroupId)
+{
+	TenantStatsHashKey *key = (TenantStatsHashKey *) palloc(sizeof(TenantStatsHashKey));
+	memset(key->tenantAttribute, 0, MAX_TENANT_ATTRIBUTE_LENGTH);
+	strlcpy(key->tenantAttribute, tenantAttribute, MAX_TENANT_ATTRIBUTE_LENGTH);
+	key->colocationGroupId = colocationGroupId;
```

Comment:
nit/food for thought: It might be nice to fill in memory from the caller instead of allocating here. Then you can store the data on the stack instead of on the heap.

```suggestion
FillTenantStatsHashKey(TenantStatsHashKey *key, char *tenantAttribute, uint32 colocationGroupId)
{
	memset(key->tenantAttribute, 0, MAX_TENANT_ATTRIBUTE_LENGTH);
	strlcpy(key->tenantAttribute, tenantAttribute, MAX_TENANT_ATTRIBUTE_LENGTH);
	key->colocationGroupId = colocationGroupId;
```